src/map/ai/controllers/pet_controller.cpp

```diff
@@ -40,7 +40,9 @@ void CPetController::Tick(time_point tick)
     TracyZoneScoped;
     TracyZoneString(PPet->getName());
 
-    if (PPet->objtype == TYPE_PET && static_cast<CPetEntity*>(PPet)->shouldDespawn(tick))
+    bool isPlayerPet = PPet->objtype == TYPE_PET || (PPet->objtype == TYPE_MOB && PPet->PMaster && PPet->PMaster->objtype == TYPE_PC);
+
+    if (isPlayerPet && static_cast<CPetEntity*>(PPet)->shouldDespawn(tick))
```

Comment:
This doesn't look particularly safe.

```cpp
if (PPet->objtype == TYPE_PET)
{
    static_cast<CPetEntity*>(PPet)->shouldDespawn(tick));
```
^ This is fine, because you've confirmed that PPet is a pet type, and then you're static casting and using it as a pet type.

By adding in this condition:
```cpp
(PPet->objtype == TYPE_MOB && PPet->PMaster && PPet->PMaster->objtype == TYPE_PC);
```

If the type is a mob, it has a master, and that master is a player - you're confirming that its a Mob AND THEN forcing it to be a Pet.

Pet is derived from Mob, sure, but `shouldDespawn` is a member function of `CPetEntity`. This is UB.

Reply:
Example:
```cpp
#include <unordered_map>
#include <vector>
#include <cmath>
#include <iostream>

enum Type
{
    Mob,
    Pet,
};

class CMobEntity
{
public:
    Type type = Type::Mob;

    virtual void sharedFunction()
    {
        std::cout << "Hello from CMobEntity\n";
    }
};

class CPetEntity final : public CMobEntity
{
public:
    Type type = Type::Pet;

    void sharedFunction() override
    {
        std::cout << "Hello from CPetEntity\n";
    }

    void specificFunction()
    {
        std::cout << "A function only CPetEntity has\n";
    }
};

int main() {
    auto* PMob = new CMobEntity();
    auto* PPet = new CPetEntity();

    if (PMob->type == Type::Mob)
    {
        std::cout << "PMob is a mob\n";
    }
    else
    {
        std::cout << "PMob is NOT a mob\n";
    }

    if (PPet->type == Type::Pet)
    {
        std::cout << "PPet is a pet\n";
    }
    else
    {
        std::cout << "PPet is NOT a pet\n";
    }

    // Try the bad casting
    if (PMob->type == Type::Mob)
    {
        auto* NotAPet = static_cast<CPetEntity*>(PMob); // This cast is UB!
        NotAPet->specificFunction(); // This might work, it might explode, we don't know!
    }

    delete PPet;
    delete PMob;

    return 0;
}
```

Reply:
As a general point, if you're dealing with entity pointers, you should always be using `dynamic_cast`, which will return `nullptr` if you try to do an invalid cast. You can also use `objtype` + `static_cast`, but only in the most simple combination.

Reply:
Yep, thanks for the comments! In the end I think the most straightforward fix is to remove the `shouldDespawn` function (as it was only called in that one spot in any case) and migrate the logic back to the pet controller. Otherwise you would need to put the function in mobentity and it is confusing to have it there given the very specific scope of the logic.

Also I changed `isSpawned` to `isAlive` because otherwise the logic ends up calling `DespawnPet` like 10 times (with a lot of warnings) when jug is dead (`DespawnPet` kills jugs) but the pet has not yet despawned.